Add image information to fluentd log entries

xperimental opened the pull request:
**\- What I did**
Add two additional fields to the log entries generated by the fluentd.

**\- How I did it**
Just a small modification to the fluentd logging driver similar to what other logging drivers do.

**\- How to verify it**
Run a docker container using the fluentd logger and observe messages with additional values:

``` bash
docker run --rm -it busybox echo "hello docker"
```

Example output (fluentd entry as JSON):

``` json
{
    "container_id": "71967c018fdb9b0cabba69caf24d9e26e9a5af8a81ee9f502ab6a7a9be4e2e66",
    "container_name": "/evil_davinci",
    "image_id": "sha256:e02e811dd08fd49e7f6032625495118e63f597eb150403d02e3238af1df240ba",
    "image_name": "busybox",
    "log": "hello docker",
    "source": "stdout"
}
```

**\- Description for the changelog**
fluentd log entries now contain `image_id` and `image_name` attributes.

**\- A picture of a cute animal (not mandatory but encouraged)**
http://hdwall-papers.com/data/wallpapers/90/WDF_1346959.jpg

xperimental committed:
Add image information to fluentd log entries.

Signed-off-by: Robert Jacob <xperimental@solidproject.de>

thaJeztah commented:
Thanks for contributing, but I'm -1 on this change; you can already add this information to log entries by specifying a custom log tag; https://docs.docker.com/engine/admin/logging/log_tags/. Doing so prevents this information to be added by _default_ (whereas it may not be useful for every user).

Is there a reason the log-tag does not suit your use case?

xperimental commented:
It is true that the log-tag can be used to get this information (and we currently have a configuration in place which does exactly that), but while building that configuration we hit a few bumps that made me think that the log-tag is not the right place for that information, at least in the context of fluentd.

The first one is that fluentd uses the log tag mostly for routing purposes and not for conveying actual content. This leads to the assumption that the tag has a certain structure in that it should be a descriptor with parts that are separated by dots and go from "less-detail" to "more-detail" from left to right. So for example you can have a tag `apache` which contains all the logs made by the apache service or you could provide more detail by having both `apache.error` and `apache.access` tags separating the error messages form the access log.

While that would still map to the docker usage of the log-tag in that you could see a tag `docker` describing all the logs coming from a docker daemon and `docker.{{.ImageName}}` those that come from a certain image it is the dotted-notation that does not quite map, because image names might get quite long. Additionally their tags can contain version numbers which have dots, which makes parsing these tags more cumbersome.

The second issue hit us when we tried to output the logs into the Google Stackdriver logging which puts additional restrictions on the log tag, in this case it did not accept any logs that contained a colon (`:`) in the log tag.

Our current solution is to use the tag to get the image name into the entry. Then we modify the log entry to have an `image_name` attribute based on the tag content and afterwards we rewrite the log tag to not contain the image name anymore to avoid errors with Stackdriver.

I do not really see a reason why the fluentd logging driver should make it so difficult for the user to get a useable message out of it.

That being said I can also see your point that there might be people not needing that information. So maybe a configuration option like the `gcp-log-cmd` one on the gcplogs driver would be a viable solution?

xperimental commented:
@thaJeztah Do you need any more information regarding this issue?

thaJeztah commented:
@xperimental I think your description is clear, but we should have a look at this. It's currently very busy preparing the upcoming 1.13 release, so there's a bit of a backlog reviewing.

Let me copy @tagomoris as well, perhaps he has more thoughts on this

xperimental commented:
Thanks for the fast response, was just worried it got lost. Take your time.

tagomoris commented:
This change looks clear and good to me too. It should helps platform management works.
But image id looks a bit long to be added in default. It will make the whole data size bigger, and may regress performance of whole logging platform.
IMO, it's better to add an option to disable it.

xperimental committed:
Make image information optional in fluentd log.

Signed-off-by: Robert Jacob <xperimental@solidproject.de>

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "fluentd-image-name" git@github.com:xperimental/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354288160
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

xperimental commented:
Sorry for the delay, but I forgot about this pull-request. I have rebased the change on the current master and added a configuration option that is off by default but can be enabled if you want to have image name & ID.

@tagomoris Could you clarify if you wanted the option just for the image ID or also for the name (like it is implemented now)? Or maybe an option where you can control if you want "nothing" (default), "only ID", "only name" or "both"?

tagomoris commented:
I prefer to have two different options to enable each keys (`fluentd-log-image-id` and `fluentd-log-image-name`). It's clear, easy to understand and robust for possible more fields.
We need some more opinion... @thaJeztah How do you think about it?

thaJeztah commented:
Wondering if this is something that will be added to other logging drivers, and if we should take a similar approach as `env` and `label`, i.e.;

```
--log-opt env=ENV1,ENV2,ENV3
--log-opt labels=label1,label2,label3
```

So;

```
--log-opt info=imageID,imageName
```

or 

```
--log-opt info=image_id,image_name
```


If we decide to go that way, should it only _add_ options, or should it _override_ the defaults, so to get the defaults _plus_ image info, should I write;

```
--log-opt info=container_name,container_id,image_id,image_name
```

Which would allow omitting `container_name` or `container_id` if desired

Maybe that's too much flexibility, just thinking about possible approaches. I'll add this to the next maintainers meeting as well, so that we can have a quick look/discussion.

xperimental commented:
> If we decide to go that way, should it only add options, or should it override the defaults [...]

I would go with "override", so that you can opt to not have the default values in there.

I'll update the PR to the current master state soon, is there any decision on which way to go with the configuration already, so I might incorporate that decision as well? I currently use fluentd and gelf loggers (in separate environments) and having most of the options "similar" is a good thing in my opinion.

I can also take a stab at implementing the configuration option change for the other loggers as well, but I guess that would be a separate PR then, right?

thaJeztah commented:
> I would go with "override", so that you can opt to not have the default values in there.
> I'll update the PR to the current master state soon, is there any decision on which way to go with the configuration already, so I might incorporate that decision as well? 

Oh, I think consensus was to go for "override", which is most flexible, and least "confusing" to the user what the end-result will be.

xperimental committed:
Add loginfo tests.

Signed-off-by: Robert Jacob <xperimental@solidproject.de>

xperimental committed:
Add new info log-opt.

Signed-off-by: Robert Jacob <xperimental@solidproject.de>

xperimental committed:
Use ID and Name as default info.

Signed-off-by: Robert Jacob <xperimental@solidproject.de>

xperimental committed:
Use info map for container information.

Signed-off-by: Robert Jacob <xperimental@solidproject.de>

xperimental committed:
Map new attribute names to old ones.

Signed-off-by: Robert Jacob <xperimental@solidproject.de>

xperimental commented:
@thaJeztah I have added an implementation for the generic "info" `--log-opt` (for now only used by the fluentd logging driver). Let me know what you think.

thaJeztah commented:
Sorry for the long delay. Design LGTM, let's move this to code review!

I see it needs a rebase, but it doesn't look bad

xperimental commented:
Noticed the need for a rebase a few days ago as well, will do that soon.

xperimental commented:
Rebase finally done, sorry for the delay.

xperimental commented:
Rebase finally done, sorry for the delay. There is a new code-path in the "extra" attributes which is not covered by the new tests yet.

cpuguy83 commented on daemon/logger/loginfo.go:
```diff
@@ -127,3 +129,45 @@ func (info *Info) ImageFullID() string {
 }
+
+// InfoMap returns information about a container as a map.
+func (info *Info) InfoMap(keyMod func(string) string) map[string]string {
```
I'm not sure I understand the purpose of this new function.
Why wouldn't we just set the keys as needed in the logger?

cpuguy83 replied:
We can always generalize this later, I'd rather just add what's being requested to fluentd and we can look at generalization if needed separately.

xperimental replied:
@cpuguy83 Just to clarify, because I might have misunderstood you the first time: Would you like me to just get rid of the `keyMod` parameter, the `InfoMap` function or "put everything in fluentd-logger implementation"?

Preparing this for a more "generalized implementation" was originally suggested by @thaJeztah ...

cpuguy83 replied:
@xperimental I'm still not sure why we need this conversion function.

xperimental replied:
You are talking about the `keyMod` function passed in, right?

cpuguy83 replied:
Yes.

xperimental replied:
The idea was that existing logger implementations might already expose some of the information available using `InfoMap` but, for historical reasons, with different key names. The `keyMod` function passed to the `InfoMap` function can be used to convert the new InfoMap name to the old one to not break the naming for existing consumers. This is already used in the fluentd logger because the new names we agreed on above do not match the existing ones for the information that can already be obtained (container ID and name).

I have not checked if the existing names match between the different loggers (because the change to the other loggers would possibly happened in a different PR), so I don't know if you want to change the new names or change fluentd's names to the new scheme. Depends on if you want to keep compatibility or want to have consistency.

xperimental committed:
Move everything back into fluentd logger.

Signed-off-by: Robert Jacob <xperimental@solidproject.de>

xperimental committed:
Make infoMap private.

Signed-off-by: Robert Jacob <xperimental@solidproject.de>

xperimental commented:
In the spirit of re-igniting the discussion I have moved everything new back into the fluentd package (except for the new tests for existing code).

cpuguy83 commented on daemon/logger/fluentd/infomap.go:
```diff
@@ -0,0 +1,48 @@
+const defaultInfoKeys = "container_id,container_name"
+
+// InfoMap returns information about a container as a map.
+func infoMap(info logger.Info) map[string]string {
```
I still don't think we should do this. We can build the map statically in `Log()`, or even when the logger is created as it is not dynamic.

cpuguy83 replied:
This all works today without this mapping.
This is going from a strong type to a weak one.
The logger still needs to know about the keys and everything. So what's the benefit of the added complexity?
This is adding more complexity, but I don't really see the benefit.

xperimental replied:
Can you be a bit more specific on what you mean with "this all works today" and where you see the additional complexity? The current approach is not configurable in any way, so some additional complexity was to be expected, I think.

I can also remove the configurable part again, if that is preferred, although it was a suggestion from your team...

cpuguy83 replied:
> Can you be a bit more specific on what you mean with "this all works today"

Loggers already use these fields today. Why do we need to change it to add a new field?

> where you see the additional complexity

This whole mapping is extra complexity.
We have structured data with predefined keys which should not be opaque to anything accessing this data.

xperimental commented on daemon/logger/fluentd/infomap.go:
```diff
@@ -0,0 +1,48 @@
+const defaultInfoKeys = "container_id,container_name"
+
+// InfoMap returns information about a container as a map.
+func infoMap(info logger.Info) map[string]string {
```
I don't think it can be more static than it is right now. It is built once per logger instance during `New()` similar to the `extra` attributes. The only improvement in this direction that comes to my mind is to combine both maps so that there is only one for loop in the `Log()` func.

vdemeester commented:
ping @xperimental @thaJeztah what's the status here ?

xperimental commented:
@vdemeester Honestly, I had forgotten that there was a new comment on this PR again after months of nothing...

The current model of feedback seems to not be working so great. Maybe it would be a better idea to meet in IRC or other synchronous medium to discuss feedback on this change and finally get it out the door.

cpuguy83 commented:
This is quite stale, I've carried this in #38312 with an implementation that doesn't add new field names (i.e. it's the same field names as used in log tag templates).

cpuguy83 commented:
Also appologies @xperimental, I clearly was not able to explain what I was hoping for until I dug into the code to look at the challenges the code base has and the requirements laid down for this feature.

cpuguy83 commented:
Going to close this because it will not be merged as is and there's been no movement on it. I also closed my PR because it still wasn't quite right.

To anyone who may want something like this, we are not against a change that enables this kind of functionality, but the implementation needs to be worked on before we can accept it.

Thanks!

xperimental commented:
Thanks @cpuguy83! I had planned to close this PR myself after you said you started your own work on this issue, but unfortunately forgot to do that.
